### PR TITLE
Add value pool/balances to non-finalized state

### LINKED ISSUE #1895: ZIP-209: Prohibit Negative Shielded Chain Value Pool Balances - Implementation
## Motivation

ZIP-209 defines how the consensus rules are altered such that blocks that produce negative shielded chain value pool balances are prohibited.

#### Transparent

Note: there is no explicit rule that the remaining balance in the transparent *chain* value pool must be non-negative. But it is implicit in the other consensus rules.

#### Sprout

> As defined in [ZIP-209], the Sprout chain value pool balance for a given block chain is the sum of all vold field
values for transactions in the block chain, minus the sum of all vnew fields values for transactions in the block chain.
>
> Consensus rule: If the Sprout chain value pool balance would become negative in the block chain created as a result of accepting a block , then all nodes MUST reject the block as invalid.

https://zips.z.cash/protocol/protocol.pdf#joinsplitbalance

#### Sapling

> As defined in [ZIP-209], the Sapling chain value pool balance for a given block chain is the negation of the sum of all valueBalanceSapling field values for transactions in the block chain.
>
> Consensus rule: If the Sapling chain value pool balance would become negative in the block chain created as a result of accepting a block , then all nodes MUST reject the block as invalid.

https://zips.z.cash/protocol/protocol.pdf#saplingbalance

#### Orchard

> Similarly to the Sapling chain value pool balance defined in [ZIP-209], the Orchard chain value pool balance for a given block chain is the negation of the sum of all valueBalanceOrchard field values for transactions in the block chain.
>
> Consensus rule: If the Orchard chain value pool balance would become negative in the block chain created as a result of accepting a block , then all nodes MUST reject the block as invalid.

https://zips.z.cash/protocol/protocol.pdf#orchardbalance

## Solution

We should implement [ZIP-209](https://zips.z.cash/zip-0209) in Zebra for the Transparent, Sprout, Sapling, and Orchard pools.

### LINKED ISSUE #2640: Write tests that check for specific value balance amounts
## Motivation

Currently, Zebra tests that value balances are valid, but we don't make sure the amounts are correct.

So the current tests would still pass if a future change accidentally skipped some or all block, transaction, or pool values.

## Related Work

PRs:
* #2381
* #2559

Tickets:
* #1895

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

PR #2652 has some bugs in the new value pool calculations that I added.

## Solution

1. Only revert the chain value pool balances when the tip is popped. Don't modify them when the root is finalized, because the balances are for the entire finalized and non-finalized chain.

2. Only update or revert the chain value pool balances once per block. (Previously, the block changes were multiplied by the number of *transactions*.)

3. Actually update the chain value balances in the non-finalized state. (Previously, the changes were ignored using `Ok(_)`.)

4. Make existing tests pass

5. Add new tests to increase value balance coverage, particularly for non-finalized chain forks, and non-finalized chain root finalization

6. Add tests that check for specific value pool balances in the finalized and non-finalized state

These changes along with PR #2652:
Closes #1895 
Closes #2640

## Review

@oxarbitrage can review this PR.

I can split it into separate PRs, if that would help with the review.

This PR is based on PR #2652 merged with a recent `main` branch.
It will need a manual rebase and squash before merging.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

**Comment by teor2345** [comment]
I've run out of time to work on this today, but here are my next steps:
1. run full mainnet and testnet syncs to tip, after deleting the buggy state
    * similar tests panicked on chain forks in PR #2652
    * 2 mainnet syncs were successful
    * 2 testnet syncs are running and should finish soon
2. fix the failing tests
    * log more debugging info when the chain tests fail with a value pool error
3. #2640
    * the bugs in this PR would have been picked up by these tests

**Comment by teor2345** [comment]
> run full mainnet and testnet syncs to tip, after deleting the buggy state
> * similar tests panicked on chain forks in PR

4 `mainnet` and `testnet` syncs were successful, so it looks like these fixes make non-finalized forks work.

**Comment by teor2345** [comment]
> > run full mainnet and testnet syncs to tip, after deleting the buggy state
> > 
> > * similar tests panicked on chain forks in PR
> 
> 4 `mainnet` and `testnet` syncs were successful, so it looks like these fixes make non-finalized forks work.

It turns out we weren't actually updating the non-finalized value pools (5ab25f4201b1457d371e435faad547a078cfb5e2).

So I'm re-running the following tests:
* run full mainnet and testnet syncs to tip
* restart synced mainnet and testnet with the latest code in this PR

(Since the finalized state value balances are correct, I don't need to delete the state.)

File: zebra-chain/src/value_balance.rs [hunk 1]
```diff
@@ -181,17 +181,17 @@ impl ValueBalance<NonNegative> {
     /// value pool.
     ///
     /// See [`Block::chain_value_pool_change`] for details.
-    pub fn update_with_block(
+    pub fn add_block(
```
Comment (oxarbitrage) [inline zebra-chain/src/value_balance.rs:184]: This name changes are a lot more clear. Thanks.

File: zebra-state/src/service/non_finalized_state/chain.rs [hunk 2]
```diff
@@ -323,39 +361,54 @@ impl Chain {
...
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+enum RevertPosition {
+    /// The chain root is being reverted via [`pop_root`],
+    /// when a block is finalized.
+    Root,
+
+    /// The chain tip is being reverted via [`pop_tip`],
+    /// when a chain is forked.
+    Tip,
+}
```
Comment (oxarbitrage) [inline zebra-state/src/service/non_finalized_state/chain.rs:379]: All the changes related to the position then reversing were a surprise to me but they make sense and seems to be well implemented. Very nice work.

**Review by oxarbitrage — APPROVED** [review]
This looks all good to me. No need to separate it in smaller PRs. I rebased into main, closed https://github.com/ZcashFoundation/zebra/pull/2652 and i am merging this one now.

File: zebra-chain/src/value_balance.rs (hunk 1 above)
Comment (teor2345) [reply zebra-chain/src/value_balance.rs:184]: Yeah, I think I changed the implementation a few weeks ago, but didn't change the names.

Sorry for the confusion!

**Pull request merged**